### PR TITLE
Phonon transport

### PR DESCRIPTION
**Pull request by pecchia**
Phonon transport tool based on libnegf and second-derivatives computed with dftb+ or other codes.
Includes dispersion computed using a supercell created with repeatgen

### CODE REVIEW

File: prog/phonons/initphonons.F90 [hunk 1]
```diff
@@ -0,0 +1,1612 @@
...
+
+  !> constants parameters
+  integer, parameter, public :: ALLMODES = 0        ! along x,y,z  
+  integer, parameter, public :: XX = 1              ! along x 
+  integer, parameter, public :: YY = 2              ! along y 
+  integer, parameter, public :: ZZ = 3              ! along z 
+  integer, parameter, public :: LONGITUDINAL = 4    ! along z
+  integer, parameter, public :: TRANSVERSE = 5      ! along x,y
+  integer, parameter, public :: INPLANE = 6         ! along x,z
+  integer, parameter, public :: OUTOFPLANE = 7      ! along y
```
Comment (awvwgk) [inline prog/phonons/initphonons.F90:161]: Using the usual enumerator style is probably better for those, occupying `xx`, `yy` and `zz` could lead to problems later.

```suggestion
  type :: TPhononEnum
    !> along x,y,z  
    integer :: ALLMODES = 0
    !> along x 
    integer :: XX = 1
    !> along y 
    integer :: YY = 2
    !> along z 
    integer :: ZZ = 3
    !> along z
    integer :: LONGITUDINAL = 4
    !> along x,y
    integer :: TRANSVERSE = 5
    !> along x,z
    integer :: INPLANE = 6
    !> along y
    integer :: OUTOFPLANE = 7
  end type TPhononEnum

  !> Possible directions
  type(TPhononEnum), parameter, public :: phononEnum = TPhononEnum()
```

File: tools/dptools/src/dptools/scripts/repeatgen.py [hunk 2]
```diff
@@ -61,6 +64,10 @@ def parse_cmdline_args(cmdlineargs=None):
             raise ScriptError(msg)
     if not (reps[0] > 0 and reps[1] > 0 and reps[2] > 0):
         raise ScriptError('Repetition numbers must be greater than zero')
+    if options.phonons:
+        if (reps[0] % 2 ==0 or reps[1] % 2 == 0 or reps[2] % 2 == 0):
```
Comment (awvwgk) [inline tools/dptools/src/dptools/scripts/repeatgen.py:68]:
```suggestion
        if (reps[0] % 2 == 0 or reps[1] % 2 == 0 or reps[2] % 2 == 0):
```

File: test/prog/phonons/Si/plotscript [hunk 3]
```diff
@@ -0,0 +1,3 @@
+set term png
+set output 'bands.png'
+plot 'phononDispersion.dat' u 1:2 pt 7 ps 1
```
Comment (awvwgk) [inline test/prog/phonons/Si/plotscript:3]: Should this get checked in to git?

File: test/prog/phonons/Si/dftb_in.hsd [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+Driver = ConjugateGradient{
+  LatticeOpt = Yes
+  FixAngles = Yes
+  MaxForceComponent = 1e-5
+}
+
+!Driver = SecondDerivatives{}
+
+Hamiltonian = DFTB {
```
Comment (awvwgk) [inline test/prog/phonons/Si/dftb_in.hsd:13]: Same as for modes, do we really need this `dftb_in.hsd` for the test?

File: test/prog/phonons/C-chain/dftb_in.hsd [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+#Driver = ConjugateGradient{
+#  LatticeOpt = Yes
+#  FixAngles = Yes
+#  FixLengths = {Yes Yes No}
+#  MaxForceComponent = 1e-5
+#}
+
+Driver = SecondDerivatives{}
+
+Hamiltonian = dftb {
```
Comment (awvwgk) [inline test/prog/phonons/C-chain/dftb_in.hsd:14]: Same here, should this file get checked in to git?

File: test/prog/phonons/C-chain-proj/dftb_in.hsd [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+#Driver = ConjugateGradient{
+#  LatticeOpt = Yes
+#  FixAngles = Yes
+#  FixLengths = {Yes Yes No}
+#  MaxForceComponent = 1e-5
+#}
+
+Driver = SecondDerivatives{}
+
+Hamiltonian = dftb {
```
Comment (awvwgk) [inline test/prog/phonons/C-chain-proj/dftb_in.hsd:14]: Also not really needed for testing `phonons`.

File: prog/phonons/make.deps [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+_modobj_initprogram = initprogram.o
+
+libnegf_int.o: $$(_modobj_ln_precision) $$(_modobj_libnegf_vars) $$(_modobj_commontypes) $$(_modobj_libnegf) $$(_modobj_lib_param) $$(_modobj_ln_allocation) $$(_modobj_ln_constants) $$(_modobj_sparsekit_drv) $$(_modobj_ln_structure) $$(_modobj_mat_conv) $$(_modobj_mpi_globals) $$(_modobj_initprogram) $$(_modobj_mat_def) $$(_modobj_ln_extract) $$(_modobj_libmpifx_module)
+libnegf_int.o = libnegf_int.o $($(_modobj_ln_precision)) $($(_modobj_libnegf_vars)) $($(_modobj_commontypes)) $($(_modobj_libnegf)) $($(_modobj_lib_param)) $($(_modobj_ln_allocation)) $($(_modobj_ln_constants)) $($(_modobj_sparsekit_drv)) $($(_modobj_ln_structure)) $($(_modobj_mat_conv)) $($(_modobj_mpi_globals)) $($(_modobj_initprogram)) $($(_modobj_mat_def)) $($(_modobj_ln_extract)) $($(_modobj_libmpifx_module))
+_modobj_libnegf_int = libnegf_int.o
```
Comment (awvwgk) [inline prog/phonons/make.deps:12]: Some relict from the make build system? Also shouldn't get on master.

File: prog/phonons/make.build [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+#                                                                              #
+#  See the LICENSE file for terms of usage and distribution.                   #
+#------------------------------------------------------------------------------#
+
+############################################################################
+#  Global variables
+############################################################################
+
+# target executables
+TARGETS = phonons
```
Comment (awvwgk) [inline prog/phonons/make.build:13]: Do we need a make build for phonons?

File: prog/phonons/masses.F90 [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+  implicit none
+  private
+
+  public getAtomicMass 
+
+  contains 
+
+  !!* Returns the atomic mass in a.m.u.
+  !!* specie is a character string 
+  function getAtomicMass(specie) result(mass)
```
Comment (awvwgk) [inline prog/phonons/masses.F90:12]: I'd suggest an interface like https://github.com/awvwgk/dftbplus/blob/atom-masses/prog/dftb%2B/lib_common/atomicmass.F90 for this purpose. This easier to maintain and also covers more cases (lowercase symbols for example).

File: test/prog/phonons/C-chain-proj/_autotest.tag [hunk 10]
```diff
@@ -0,0 +1,28 @@
+transmission        :real:2:81,1
```
Comment (aradi) [inline test/prog/phonons/C-chain-proj/_autotest.tag:1]: Is this really tested? (Would program `phonons` be executed during the tests?)

File: tools/dptools/src/dptools/scripts/repeatgen.py [hunk 11]
```diff
@@ -125,3 +136,35 @@ def _repeatgeo(geo, latvecs, repeats):
     newgeo = Geometry(geo.specienames, allinds, allcoords, latvecs=newlatvecs,
                       origin=geo.origin)
     return newgeo
+
+def _repeatgeo2(geo, latvecs, repeats):
```
Comment (aradi) [inline tools/dptools/src/dptools/scripts/repeatgen.py:140]: maybe we can merge this into the original routine...

Comment (pecchia) [reply tools/dptools/src/dptools/scripts/repeatgen.py:140]: _repeatgeo and _repeatgeo2 can be merged. 
Is it fine to pass option.phonons to the subroutine ?

Comment (pecchia) [reply tools/dptools/src/dptools/scripts/repeatgen.py:140]: there is one small issue: in repeatgen the loops are in order [0], [1], [2], whereas in geo2 they are in reverse, [2],[1],[0]. So I'll have to duplicate the core of the subroutine. 
Still like this ?

**Comment by pecchia** [comment]
Hopefully solved the problem. I merged with master and aligned all submodules with new cmake system. 
Maybe we should write a simple script to align things automatically (e.g. on all submodules do a git co master; git pull).  
It does compile and works on dp-devels with ifort-19.

**Comment by pecchia** [comment]
I am puzzled. Do not understand what is failing on Travis and on dp-devels. 
Apparently Nag7 and intel19-devel fail randomly either or both phonons/C-chain tests. 
BUT all tests pass if I compile manually on dp-devel.

**Comment by aradi** [comment]
I've squashed the long history into two commits, see #672. If you agree, I'd propose to merge that instead of this PR.

**Comment by pecchia** [comment]
Yes , it is fine to squash all-

 

Concerning #669, I do not understand well. When I merged master into branch I had those changes in scalapackfx.

 

The branch libnegf_up contains a small change in initprogram.F90 on the dftb+ side and more changes in libNEGF.

Basically updating the submodule reference to current libNEGF HEAD. 

 

I can restart from a clean new branch from master. But then merge phonon first, so I can co that.

Thanks,

 

 

Alex

 

 

 

Da: Bálint Aradi [mailto:notifications@github.com] 
Inviato: mercoledì 11 novembre 2020 13:39
A: dftbplus/dftbplus
Cc: Alessandro Pecchia; Author
Oggetto: Re: [dftbplus/dftbplus] Phonon transport (#516)

 

I've squashed the long history into two commits, see #672 <https://github.com/dftbplus/dftbplus/pull/672> . If you agree, I'd propose to merge that instead of this PR.

**Comment by aradi** [comment]
#672 had been merged with the same content.

**Pull request closed without merging**